scripts/build.py

```diff
@@ -630,52 +748,128 @@ def rebuild_new_files(
         target_file_path = os.path.join(target_path, file_name)
         ensure_directory_exists(target_file_path)
-        minify_func(source_file_path, target_file_path, file_hashes, file_name)
+        task = threading.Thread(target=minify_func, args=(
+            source_file_path, target_file_path, file_hashes, file_name))
+        build_tasks.append(task)
+    return build_tasks
 
 
-def remove_deleted_files(dev_directory, staging_directory):
-    """Walk the staging directory and remove files that have since been removed
-    from the DEV directory.
+def generate_build_task_to_build_directory(dirnames_dict, file_hashes):
+    """Queue up build tasks to build all files in source directory if there is
+    no existing staging directory. Otherwise, selectively queue up build task
+    to build recently changed files and save them at staging directory.
 
     Args:
-        dev_directory: str. Path relative to /oppia where DEV files are located.
+        dirnames_dict: dict(str, str, str). A dict containing directory paths to
+            the directory containing source files to be built, the staging
+            directory and the final directory containing built files.
+        file_hashes: dict(str, str). Dictionary with filepaths as keys and
+        hashes of file content as values.
+
+    Returns:
+        build_tasks: deque(Thread). A deque that contains all build tasks queued
+            to be processed.
+    """
+    source_dir = dirnames_dict['dev_dir']
+    staging_dir = dirnames_dict['staging_dir']
+    out_dir = dirnames_dict['out_dir']
+    build_tasks = collections.deque()
+    delete_tasks = collections.deque()
+    if not os.path.isdir(staging_dir):
+        # If there is no staging dir, perform build process on all files.
+        print 'Creating new %s folder' % staging_dir
+        build_tasks.append(generate_build_task_to_build_files(
+            source_dir, staging_dir, file_hashes))
+    else:
+        # If staging dir exists, rebuild all HTML and Python files.
+        mandatory_file_formats = ('.html', '.py')
+        print (
+            'Staging dir exists, re-building all %s files'
+            % str(mandatory_file_formats))
+        build_tasks += generate_build_task_to_build_files(
+            source_dir, staging_dir, file_hashes,
+            file_formats=mandatory_file_formats)
+        # Compare source files with already built files using file hashes.
+        dev_dir_hashes = get_file_hashes(source_dir)
+        print 'Getting files that have changed between %s and %s' % (
+            source_dir, out_dir)
+        recently_changed_filenames = get_recently_changed_filenames(
+            dev_dir_hashes, out_dir)
+        if recently_changed_filenames:
+            # Only re-build files that have changed since last build.
+            print 'Re-building recently changed files at %s' % source_dir
+            build_tasks += generate_build_task_to_rebuild_new_files(
```

Comment:
So, above, I'm seeing generate_build_task_to_build_files() on line 801, and this looks like it's doing something similar. If they're doing something similar, why the duplication? If not, perhaps make that clearer in the naming.

Reply:
So it does look like some duplication is going for me. I model the `build` function to build from different sources:
build_from_directory (wholesale file building, nothing left behind)
build_from_filepaths(only build listed filepaths)